common/scala/src/main/scala/org/apache/openwhisk/core/yarn/YARNContainerFactory.scala

```diff
@@ -0,0 +1,476 @@
+    unusedcpuShares: Int)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+
+    logging.info(this, s"Using YARN to create a container with image ${actionImage.name}...")
+
+    Future {
+      serviceLock.synchronized {
+        val newContainerCount = containerCounts.getOrElse(actionImage.name, 0) + 1
+        val body = "{\"number_of_containers\":" + newContainerCount + "}"
+        val response = submitRequestWithAuth(
+          HttpMethods.PUT,
+          s"${yarnConfig.masterUrl}/app/v1/services/${yarnConfig.serviceName}/components/${actionImage.name}",
+          body)
+        response match {
+          case httpresponse(StatusCodes.OK, content) =>
+            logging.info(this, s"Added container: ${actionImage.name}. Response: $content")
+            containerCounts(actionImage.name) = newContainerCount
+
+          case httpresponse(_, _) => handleYARNRESTError(response)
+        }
+
+        //Wait for container to come up
+        val serviceDef = waitForContainerStability(actionImage)
+
+        //Update container list with new container details
+        val updatedContainerList = serviceDef.components
+          .filter(c => c.name.equals(actionImage.name))
+          .flatMap(c => c.containers.getOrElse(List[ContainerDefinition]()))
+        logging.info(this, s"YARN service now has ${updatedContainerList.size} total containers")
+
+        val newContainers = updatedContainerList.filter(a =>
+          !containerMap.getOrElse(actionImage.name, List[ContainerDefinition]()).contains(a))
+
+        //Synchronized block should prevent this from happening
+        if (newContainers.length > 1)
+          logging.warn(this, "More than one new container...")
+
+        containerMap(actionImage.name) = updatedContainerList
+        val newContainer = newContainers.head
+
+        val containerAddress = ContainerAddress(newContainer.ip.getOrElse("127.0.0.1")) //port 8080 is default
+        val containerId = ContainerId(newContainer.id)
+
+        logging.info(this, s"New Container: ${newContainer.id}, $containerAddress")
+        new YARNTask(containerId, containerAddress, ec, logging, as, actionImage, yarnConfig, this)
+      }
+    }
+  }
+
+  def waitForServiceStart(): Unit = {
+    var started = false
+    var retryCount = 0
+    val maxRetryCount = serviceStartTimeoutMS / retryWaitMS
+    while (!started && retryCount < maxRetryCount) {
+      val serviceDef = downloadServiceDefinition()
+
+      //Should never happen
+      if (serviceDef == null)
+        throw new IllegalArgumentException("Service not found?")
+
+      serviceDef.state match {
+        case "STABLE" | "STARTED" =>
+          logging.info(this, "YARN service achieved stable state")
+          started = true
+
+        case state =>
+          logging.info(
+            this,
+            s"YARN service is not in stable state yet ($retryCount/$maxRetryCount). Current state: $state")
+          Thread.sleep(retryWaitMS)
```

Comment:
this sleep here inside the future should be avoided.
as a brute force mechanism, you can wrap the method with `blocking {}`.